### DISCUSSION TITLE
Trying to add a turtle upgrade (1.19.2)

### DISCUSSION THREAD

**Discussion by blockninja124** [cc-tweaked/CC-Tweaked#1831] [Q&A] [answered]
I'm trying to make a CC:tweaked addon for 1.19.2 (1.101.4) with a custom peripheral. I mostly want this peripheral to be a turtle, but I'm fine with having a hidden peripheral block for now. I've managed to add my custom peripheral with the Lua functions and a block entity, but I am having trouble making it a turtle upgrade. From the docs [here](https://tweaked.cc/mc-1.20.x/javadoc/dan200/computercraft/api/turtle/ITurtleUpgrade.html) (which is annoyingly for 1.20.x), I know that I have to have some kinda of serialiser registered to the event bus that somehow adds my turtle upgrade class, but I'm really stumped at how to implement this. The serialiser example on the docs is unclear on where the code would be, and also uses `TurtleUpgradeSerialiser.TYPE` which doesn't seem to exist? Apologies if this is just me not understanding Java/forge stuff very well yet. Any help at all would be greatly appreciated, even just a nudge in the right direction!

**Comment by SquidDev** [comment, marked as answer]
Ahh yes, the javadocs default to 1.20.1, but it should be possible to get the 1.19 version by [changing the version in the URL](https://tweaked.cc/mc-1.19.x/javadoc/dan200/computercraft/api/turtle/TurtleUpgradeSerialiser.html). Though that is for 1.19.4 rather than 1.19.2, so it may be more accurate just to [consult the source](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.2/src/main/java/dan200/computercraft/api/turtle/TurtleUpgradeSerialiser.java). Sorry, it would be nice to have a version switcher, but have never managed to figure out how to wire one up.

> but I'm really stumped at how to implement this.

The registration code should just live in the same place that you handle registration for blocks and items. But rather than adding it to the blocks registry, you need to add the serialiser to the upgrade serialiser registry.

> also uses TurtleUpgradeSerialiser.TYPE

Ahh, sorry! It looks like a field was renamed, but the example wasn't. That should be `TurtleUpgradeSerialiser.REGISTRY_ID` instead.

**Reply by blockninja124** [reply]
Thank you so much for your help! I managed to get my peripheral on a turtle!
